File: 09.Greed贪心/leetcode_3720_lexGreaterPermutation.cpp
#include <iostream>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    string lexGreaterPermutation(string s, string target) {
        unordered_map<int, int> left;
        for (auto& c : s){
            left[c]++;
        }
        for (auto& c : target){
            left[c]--;
        }

        for (int i=s.size()-1; i>=0; i--){
            left[target[i]] += 1;
            bool flag = true;
            for (auto& kv : left){
                if (kv.second < 0){
                    flag = false;
                }
            }
            if (!flag){
                continue;
            }

            for (char c=target[i]+1; c<='z'; c++){
                if (left[c] == 0){
                    continue;
                }

                left[c] -= 1;
                string res = target.substr(0, i);
                res += c;

                for (char cur='a'; cur<='z'; cur++){
                    if (left[cur] > 0){
                        res += string(left[cur], cur);
                    }
                }
                return res;
            }
        }
        return "";
    }
};

int main(){
    Solution sl;
    string s = "abc", target = "bba";
    cout << sl.lexGreaterPermutation(s, target) << endl;
}

File: 01.DFS递归/leetcode_3720_lexGreaterPermutation.py
from functools import cache
from collections import Counter

class Solution:
    def lexGreaterPermutation(self, s: str, target: str) -> str:
        maxest = "".join(sorted(list(s), reverse = True))
        if maxest <= target:
            return ""
        n = len(s)
        cnt = Counter(s)
        @cache
        def dfs(indx, cur, flag):
            if flag == True:
                leftS = ""
                for k,v in cnt.items():
                    leftS += k*v
                return cur+"".join(sorted(list(leftS)))
            if indx == n:
                if cur > target:
                    return cur
                return "{"*n
            ans = '{'*n
            for k in cnt.keys():
                if (k < target[indx] and flag == False) or cnt[k] == 0:
                    continue
                cnt[k] -= 1
                if k == target[indx]:
                    ans = min(ans, dfs(indx+1, cur+k, flag))
                else:
                    ans = min(ans, dfs(indx+1, cur+k, True))
                cnt[k] += 1
            return ans
        res = dfs(0, "", False)
        return "" if res=='{'*n else res

if __name__ == "__main__":
    sl = Solution()
    s, target = "abc", "bba"
    print(sl.lexGreaterPermutation(s, target))
